### PR TITLE
DDF-1827: Create Migratable Utility module

### PR DESCRIPTION
**Pull request by lessarderic**
Reviewers:
@andrewkfiedler 
@figliold
@garrettfreibott 
@oconnormi 
@Lambeaux

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/codice/ddf/pull/501" target="_blank">ReviewNinja

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Utility class for services that implement {@link org.codice.ddf.migration.Migratable}.
+ * <p/>
+ * <p>
+ * <b>This code is experimental. While this interface is functional and tested, it may change or be
+ * removed in a future version of the library.</b>
+ * </p>
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Do we really want to say that this is experimental?

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+ */
+public class MigratableUtil {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MigratableUtil.class);
+
+    private static final String UNREAL_PATH_WARNING =
+            "The value for property [%s] is set to a path [%s] that could not coerced into a real path; "
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Should this say "...not be coerced..."?

Comment (garrettfreibott) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: There's some sort of conspiracy going on. Phil had the same typo that he fixed in his original symbolic link PR.

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+     * @param systemProperty  name of the {@link System} property that contains the path to the
+     *                        source file
+     * @param exportDirectory root directory where the file will be copied. If the file'
+     *                        relative path included any directories, those will be re-created
+     *                        under this directory.
+     * @param warnings        any warnings generated during this operation (e.g., source file outside
+     *                        of {@code ddf.home}) will be added to this collection
+     */
+    public void copyFileFromSystemPropertyValue(@NotNull String systemProperty,
+            @NotNull Path exportDirectory, @NotNull Collection<MigrationWarning> warnings) {
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Do you think we should put throws MigrationException on this?

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+     *                         source file to copy
+     * @param javaProperty     name of the property inside the Java properties file that contains
+     *                         the path to the source file
+     * @param exportDirectory  path to the destination
+     * @param warnings         any warnings generated during this operation (e.g., source file
+     *                         outside of {@code ddf.home}) will be added to this collection
+     */
+    public void copyFileFromJavaPropertyValue(@NotNull Path propertyFilePath,
+            @NotNull String javaProperty, @NotNull Path exportDirectory,
+            @NotNull Collection<MigrationWarning> warnings) {
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Do you think we should put throws MigrationException on this?

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        notNull(propertyFilePath, "Java properties file cannot be null");
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Don't we need the other notNull() checks?

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 3 above)
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Don't we need the other notNull() checks?

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+        return prop;
+    }
+
+    private void copy(Path sourceFile, Path exportDirectory, Collection<MigrationWarning> warnings,
+            BooleanSupplier isSourceMigratable) {
+        notNull(sourceFile, "Source file cannot be null");
+        notNull(exportDirectory, "Destination cannot be null");
+        notNull(warnings, "Warning collection cannot be null");
+
```
Comment (figliold) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Do you think it is better to have the notNull() checks in the methods that have the @NotNull annotations?

File: platform/migration/pom.xml [hunk 7]
```diff
@@ -22,8 +22,32 @@
...
     <name>DDF Migration</name>
+
+    <dependencies>
+<!--
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>1.1.0.Final</version>
+        </dependency>
+-->
```
Comment (figliold) [inline platform/migration/pom.xml]: Can we remove this commented out dependency?

**Comment by oconnormi** [comment]
:+1:

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 1 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Until all of the migration stuff is done, including how we're going to handle async operations, I'd rather put that on our public facing classes and interfaces. For instance, if this all becomes asynchronous, we may want to get rid of the `Collection<MigrationWarning>` parameter and have the operations log things some other way.

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 2 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Good catch guys! Looks like I missed one possible error path in my code, i.e., `getRealPath()` fails which means that the path doesn't exist or can't be accessed. I might change the warning to something a little bit less cryptic though, e.g., `The path [%s] does not exist or cannot be read;`. Thoughts?

Comment (oconnormi) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Sounds good to me

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 3 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Added `MigratableException` on all the methods that didn't have it.

Regarding the `notNull()` checks, the private `copy` method already does that and I don't think we want them twice.

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 4 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Added.

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 5 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: Done in the private `copy()` method.

File: platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java (hunk 6 above)
Comment (lessarderic) [inline platform/migration/platform-migratable-util/src/main/java/org/codice/ddf/migration/util/MigratableUtil.java]: I guess we could, but do we want to repeat all those checks everywhere when we have a common private method that already does it?

File: platform/migration/pom.xml (hunk 7 above)
Comment (lessarderic) [inline platform/migration/pom.xml]: Removed.

**Comment by oconnormi** [comment]
Heroed successfully :+1:

**Pull request merged**